e2e: set appropriate groups in SAR of kubelet authz test

bertinatto opened the pull request:
#### What type of PR is this?

/kind bug
/kind failing-test

#### What this PR does / why we need it:

This PR sets the appropriate groups (`system:authenticated`, `system:serviceaccounts`, `system:serviceaccounts:<ns>`) in the SAR used in `WaitForAuthzUpdate()` to accurately simulate the service account’s identity.

#### Which issue(s) this PR fixes:

Fixes #

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

bertinatto commented:
/retest

bertinatto commented:
/assign @rphillips @vinayakankugoyal

SergeyKanzhelev commented:
/assign @tallclair

tallclair commented on test/e2e/node/kubelet_authz.go:
```diff
@@ -42,24 +43,24 @@ var _ = SIGDescribe(feature.KubeletFineGrainedAuthz, func() {
 	ns := f.Namespace.Name
-	saName := authzSubresource
-	crName := authzSubresource
+	saName := authzSubresource + "-sa-" + rand.String(5)
+	crName := authzSubresource + "-cr-" + rand.String(5)
```
optional: use `GenerateName` instead

bertinatto replied:
I remove the commit changing this in favor of https://github.com/kubernetes/kubernetes/pull/131249

tallclair commented on test/e2e/node/kubelet_authz.go:
```diff
@@ -42,24 +43,24 @@ var _ = SIGDescribe(feature.KubeletFineGrainedAuthz, func() {
-	saName := authzSubresource
-	crName := authzSubresource
+	saName := authzSubresource + "-sa-" + rand.String(5)
+	crName := authzSubresource + "-cr-" + rand.String(5)
```
Is the ClusterRole and ClusterRoleBinding ever cleaned up?

tallclair commented on test/e2e/node/kubelet_authz.go:
```diff
@@ -112,7 +113,12 @@ func runKubeletAuthzTest(ctx context.Context, f *framework.Framework, endpoint,
 	)
-	framework.ExpectNoError(err)
+
+	if err != nil && failOnAuthzUpdateErr {
+		framework.ExpectNoError(err)
+	} else if err != nil {
+		framework.Logf("Warning: Failed waiting for authorization: %v", err)
+	}
```
This is ignoring all errors, not just time out errors.

bertinatto replied:
This is gone

tallclair reviewed (commented):
I don't think I agree with the direction of this. I don't think we want to just ignore these errors across all platforms because of an openshift-specific authorization configuration. It would be better to skip this specific test case in that test environment. If you need to add a label to the test make filtering easier that's ok.

bertinatto committed:
test/e2e/node/kubelet_authz.go: fix SAR to include service account groups

This sets the appropriate groups (system:authenticated, system:serviceaccounts,
system:serviceaccounts:<ns>) in the SAR used in WaitForAuthzUpdate to accurately
simulate the service account’s identity.

bertinatto commented:
/reopen

vinayakankugoyal commented on test/e2e/framework/auth/helpers.go:
```diff
@@ -45,11 +45,12 @@ type bindingsGetter interface {
 // WaitForAuthzUpdate checks if the give user can perform named verb and action
 // on a resource or subresource.
-func WaitForAuthzUpdate(ctx context.Context, c v1authorization.SubjectAccessReviewsGetter, user string, ra *authorizationv1.ResourceAttributes, allowed bool) error {
+func WaitForAuthzUpdate(ctx context.Context, c v1authorization.SubjectAccessReviewsGetter, user string, groups []string, ra *authorizationv1.ResourceAttributes, allowed bool) error {
```
just a thought: maybe expecting everyone to add the right groups is too much?

Could we in this function detect that the user had prefix system:serviceaccount: and add those groups automatically.

I don't feel strongly about this so feel free to ignore this suggestion.

bertinatto replied:
IMO we should give callers the opportunity to choose whether to include that group or not. I don't feel strong about it as well.

vinayakankugoyal replied:
Makes sense. Thanks!

tallclair commented on test/e2e/node/kubelet_authz.go:
```diff
@@ -104,6 +104,11 @@ func runKubeletAuthzTest(ctx context.Context, f *framework.Framework, endpoint,
 		serviceaccount.MakeUsername(ns, saName),
+		[]string{
+			"system:authenticated",
+			"system:serviceaccounts",
```
Better to use `serviceaccount.MakeGroups(ns)` instead, but that doesn't include system:authenticated. Do you need that group? If so, maybe it can be include in `MakeGroups`?

bertinatto replied:
Fixed to use `serviceaccount.MakeGroups(ns)` and manually added `system:authenticated`. It doesn't seem like that group fits well in the return value of `serviceaccount.MakeGroups()`, as that function seems to be related to SA groups only.

Unfortunately I do need that group, and it seems reasonable to add it here as it present in real requests, since it's added in the authentication layer and is always present when the user is authenticated:

https://github.com/kubernetes/kubernetes/blob/e36a168a63039ff76e659a7ea2a91798d65c9f82/staging/src/k8s.io/apiserver/pkg/authentication/group/authenticated_group_adder.go#L56

haircommander commented:
/cc

haircommander commented:
/triage accepted

tallclair commented:
/lgtm
/approve

bertinatto commented:
/retest

bertinatto commented:
/assign @pohly

pohly reviewed (approved):
/approve

bertinatto commented:
/retest

bertinatto commented:
/test pull-kubernetes-e2e-kind-ipv6

kevinrizza commented:
/test pull-kubernetes-e2e-kind-ipv6
